File: SearchIn2DArray.cpp
#include <bits/stdc++.h>

using namespace std;

/* We are doing Binary search here.
 * Time Complexity   : O(log(n)) - Search space is divided by half every iteration
 * Space Complexity  : O(1)      - No extra space is being used 
 */
bool searchMatrix(vector<vector<int>>& matrix, int target)
{
    int rows = matrix.size(), cols = matrix[0].size();
    int low = 0, high = (rows * cols) - 1;
    int mid, row, col;
    while (low <= high) {
        mid = (low + (high - low) / 2);
        row = mid / cols;
        col = mid % cols;
    
        if (matrix[row][col] == target) {
            return true;
        } else if (matrix[row][col] > target) {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
        
    return false;
}

int main (int argc, char *argv[])
{
    vector<vector<int>> m =  {{1,3,5,7},{10,11,16,20},{23,30,34,60}};
    if (argc != 2) {
        cout << "USAGE: " << argv[0] << " <number>" << endl;
        return 0;
    }

    if (searchMatrix(m, atoi(argv[1]))) {
        cout <<argv[1] << " is present in the Matrix " << endl;
    } else {
        cout <<argv[1] << " is not present in the Matrix " << endl;
    }

    return 0;
}

File: Search_in_2DArray.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)-1
        n = len(matrix[0])-1
        left = 0
        right = n
        if m == 0:
            left = 0
            right = n
            while left <= right:
                mid = (left+right)//2
                if matrix[0][mid] == target:
                    return True
                elif matrix[0][mid]>target:
                    right = mid-1
                else:
                    left = mid+1
            return False
        elif n == 0:
            left = 0
            right = m
            while left <= right:
                mid = (left+right)//2
                if matrix[mid][0] == target:
                    return True
                elif matrix[mid][0]>target:
                    right = mid-1
                else:
                    left = mid+1
            return False
        else:
            left = 0
            right = m
            while left <= right:
                mid = (left+right)//2
                if matrix[mid][n] == target:
                    return True
                elif matrix[mid][n]>target and matrix[mid][0]<= target:
                    break;
                elif matrix[mid][n]>target:
                    right = mid-1
                else:
                    left = mid + 1
            if left > right:
                return False
            left = 0
            right = n
            while left <= right:
                mid1 = (left+right)//2
                if matrix[mid][mid1] == target:
                    return True
                elif matrix[mid][mid1] > target:
                    right = mid1 - 1
                else:
                    left = mid1 + 1
            return False
            
            
        